Training/Training/file.c

```diff
@@ -1,13 +1,169 @@
+         if (arr[j] > arr[j + 1]) {
+            arr[j] = arr[j] + arr[j + 1];
+            arr[j + 1] = arr[j] - arr[j + 1];
+            arr[j] = arr[j] - arr[j + 1];
+         }
+      }
+   }
+}
+
+int BinarySearch (int sortArr[], int key, int size) {
+   int low = 0, high = size - 1, result = -1;
+   while (low <= high) {
+      int mid = low + (high - low) / 2;
+      if (sortArr[mid] == key) {
+         result = mid;
+         high = mid - 1;
+      }
+      if (sortArr[mid] < key) low = mid + 1;
+      else high = mid - 1;
+   }
+   return result;
+}
+
+int GetUserInput (char* prompt) {
+   int64_t number = 0;
+   for (;;) {
+      char strNum[MAX_DIGIT];
+      char* endPtr = NULL;
+      printf ("%s", prompt);
+      char* strPtr = fgets (strNum, sizeof (strNum), stdin);
+      char* validInput = strchr (strNum, '\n');
+      number = strtoll (strNum, &endPtr, 10);
+      if (validInput == NULL)  while (getchar () != '\n');
+      if (strPtr == NULL || validInput == NULL || *endPtr != '\n' || (isdigit (strNum[0]) == 0 && strNum[0] != '-') || number < INT32_MIN || number > INT32_MAX) {
+         printf (ANSI_COLOR_RED"Invalid Number!!! Please re-enter\n"ANSI_RESET_ALL);
+         continue;
+      }
+      break;
+   }
+   return (int)number;
+}
+
+void SearchAndSort () {
+   int size = 0;
+   do {
+      size = GetUserInput ("\nEnter the size of array:");
+      if (size <= 0)  printf (ANSI_COLOR_RED"Invalid Number!!! Please re-enter\n"ANSI_RESET_ALL);
+   } while (size <= 0);
+   int* array = (int*)malloc (size * sizeof (int));
+   if (array == NULL) {
+      printf ("Memory allocation failed!\n");
+      return;
+   }
+   printf ("\nEnter the elements of the array:\n");
+   for (int i = 0; i < size; i++) {
+      printf ("[%d]", i);
+      array[i] = GetUserInput ("");
+   }
+   BubbleSort (array, size);
+   printf ("Sorted Array: ");
+   for (int i = 0; i < size; i++) printf (ANSI_COLOR_YELLOW"%d "ANSI_RESET_ALL, array[i]);
+   int index = BinarySearch (array, GetUserInput ("\n\nWhich element's index do you want to find?\n"), size);
+   printf ((index == -1) ? ANSI_COLOR_GREEN"Element is not present in array\n\n"ANSI_RESET_ALL : ANSI_COLOR_GREEN"Element is present at index %d\n\n"ANSI_RESET_ALL, index);
+   free (array);
+}
+
+void TestCases () {
+   printf (ANSI_COLOR_YELLOW"\n-----------------TESTCASES-----------------\n"ANSI_RESET_ALL);
+   int inpArray[][TEST_SIZE] = { { 25,20,-1,-8,0,2 },{ -214748,2547,154,87,4,9,10 } ,{ -3,25,-9,-5,4,1,-21 } ,{ 0,1,2,3 } ,
+                         { 0,-1,-2,-3 },{ 2000,-2147483647,800,-3,-7 } ,{ 17,17,17,17,17 },{ 1,12,5,5,65,5 } };
+   int outArray[][TEST_SIZE] = { { -8,-1,0,2,20,25 } ,{ -214748,4,9,10,87,154,2547 } , { -21,-9,-5,-3,1,4,25 } ,{ 0,1,2,3 } ,
+                         { -3,-2,-1,0 },{ -2147483647,-7,-3,800,2000 },{ 17,17,17,17,17 },{ 1,5,5,5,12,65 } };
+   int arrSize = sizeof (inpArray) / sizeof (inpArray[0]);
+   int sizes[] = { 6,7,7,4,4,5,5,6 };
```

Comment:
Instead of storing the size of the inputs as an array it can be calculated inside the for loop and stored in a local variable and used. 

Alternate approach can be you can use an array as used above and store its value in a local variable inside the loop and use it inside instead of accessing it each time from array.